### DISCUSSION TITLE
Method declaration line number

### DISCUSSION THREAD

**Discussion by freya022** [classgraph/classgraph#695] [Ideas]
Hi,

### About the use case
As I am developing a rather reflection-heavy framework, I often need to throw exceptions in case the user didn't declare a method correctly, or if they used an incorrect annotation for example.

When making these errors, even if I put the faulty method's signature in the exception, it still takes some time for the user, including myself, to find back the faulty class.

### The use case solution
The solution would be to print a text similar to when an exception happens, it is formatted in such a way that IDEs recognize them and make an hyperlink to the method declaration. (such as `System.err.println("MyCommand.declare(MyCommand.kt:63)")`)

Unfortunately I could not find a practical way that would let me get the line at which a method is declared, I could only find the [JVM specification about it](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.7.12)

As it is a rather specific suggestion, it is fine if its not implemented, but I would appreciate having better exceptions thanks to this feature.

**Comment by lukehutch** [comment]
Thanks for the suggestion -- implemented, see 04457b9061424d8889003037302fc9a0bdc217c8.

Works for me, but YMMV -- I don't know under which circumstances the line numbers are stripped from classfiles, but they are not required to be there.

The methods you want are `MethodInfo#getMinLineNum` and `MethodInfo#getMaxLineNum`. The minimum line number is the first non-empty line in the body of the function, not the line the method is declared at. Hopefully this is good enough for what you need. I don't know how to get more fine-grained than that.

Also, I don't see any similar line number info for fields (unfortunately)...

Released in 4.8.149.

**Reply by freya022** [reply]
Hi, thanks a lot for the quick response, I just tried out the feature and it seems like it works pretty well, the first line of the function body is perfectly fine, i've checked with the bytecode and it completely corresponds.

It's fine if the info isn't there for fields, I can see the line number in the bytecode but it doesn't seem be exactly linked to the field ¯\\_(ツ)_/¯

![image](https://user-images.githubusercontent.com/41875020/177549147-3162e3f1-5317-48e9-95d8-2c26aea4b821.png)

**Comment by lukehutch** [comment]
Glad it's (mostly) working for you!

Where do you see the line number in the bytecode for fields?

**Reply by freya022** [reply]
Using the kotlin decompiler:
![image](https://user-images.githubusercontent.com/41875020/177553026-ffd23040-cb0e-48c0-b732-dfa67107c634.png)

Though to be clear, i don't need the field line number

**Reply by freya022** [reply]
I just looked over the bytecode of a field that is initialized in the constructor, and the line number of the field is missing, so probably the line number was here earlier because the initialization was on the same line.

**Comment by lukehutch** [comment]
Oh, if the line numbers are in the bytecode, then there's no way for ClassGraph to get those numbers. ClassGraph will never parse bytecode, only class metadata.
